vms/platformvm/state/state.go

```diff
@@ -2220,3 +2340,181 @@ func parseStoredBlock(blkBytes []byte) (blocks.Block, choices.Status, bool, erro
 
 	return blkState.Blk, blkState.Status, true, nil
 }
+
+func (s *state) PruneAndIndex(lock sync.Locker, log logging.Logger) error {
+	lock.Lock()
+	shouldPrune, err := s.shouldPrune()
+	if err != nil {
+		lock.Unlock()
+		return fmt.Errorf(
+			"failed to check if the database should be pruned: %w",
+			err,
+		)
+	}
+	if !shouldPrune {
+		lock.Unlock()
+		return nil
+	}
+
+	// It is possible that new blocks are added after grabbing this iterator. New
+	// blocks are guaranteed to be accepted and height-indexed, so we don't need to
+	// check them.
+	blockIterator := s.blockDB.NewIterator()
+	// Releasing is done using a closure to ensure that updating blockIterator will
+	// result in having the most recent iterator released when executing the
+	// deferred function.
+	defer func() {
+		blockIterator.Release()
+	}()
+
+	// While we are pruning the disk, we disable caching of the data we are
+	// modifying. Caching is re-enabled when pruning finishes.
+	//
+	// Note: If an unexpected error occurs the caches are never re-enabled.
+	// That's fine as the node is going to be in an unhealthy state regardless.
+	oldBlockIDCache := s.blockIDCache
+	s.blockIDCache = &cache.Empty[uint64, ids.ID]{}
+	lock.Unlock()
+
+	log.Info("starting state pruning and indexing")
+
+	blkBytes := blockIterator.Value()
```

Comment:
`.Value()` should always be called after `.Next()`. In the prior PR - we did some work to get the start progress (because we were deleting everything). In this case `startProgress` is only going to go up to the first `Accepted` block... So I feel like we can safely assume that `startProgress` is 0 every time (which should make the eta code pretty easy)

Reply:
cleaned up 👍🏾